extensions/amp-consent/0.1/cookie-writer.js

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BASE_CID_MAX_AGE_MILLIS} from '../../../src/service/cid-impl';
+import {ConsentLinkerReader} from './linker-reader';
+import {Services} from '../../../src/services';
+import {getMode} from '../../../src/mode';
+import {hasOwn} from '../../../src/utils/object';
+import {isInFie} from '../../../src/iframe-helper';
+import {isObject} from '../../../src/types';
+import {isProxyOrigin} from '../../../src/url';
+import {setCookie} from '../../../src/cookies';
+import {user} from '../../../src/log';
+
+const TAG = 'amp-consent/cookie-writer';
+
+const RESERVED_KEYS = {
+  'referrerDomains': true,
+  'enabled': true,
+  'cookiePath': true,
+  'cookieMaxAge': true,
+  'cookieSecure': true,
+  'cookieDomain': true,
+};
+
+/** @const @type {!Object<string, boolean>} */
+const CONSENT_COOKIE_WRITE_VARS_ALLOWED_LIST = {
+  'LINKER_PARAM': true,
+};
+
+export class CookieWriter {
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   * @param {!JsonObject} config
+   */
+  constructor(win, element, config) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.element_ = element;
+
+    /** @private {?Promise} */
+    this.writePromise_ = null;
+
+    /** @private {!JsonObject} */
+    this.config_ = config;
+
+    this.linkerReader_ = new ConsentLinkerReader(win);
+  }
+
+  /**
+   * @return {!Promise}
+   */
+  write() {
+    if (!this.writePromise_) {
+      this.writePromise_ = this.init_();
+    }
+    return this.writePromise_;
+  }
+
+  /**
+   * Parse the config and write to cookie
+   * Config looks like
+   * cookies: {
+   *   enabled: true/false, //Default to true
+   *   cookieNameA: {
+   *     value: cookieValueA (LINKER_PARAM)
+   *   },
+   *   cookieValueB: {
+   *     value: cookieValueB
+   *   }
+   *   ...
+   * }
+   * @return {!Promise}
+   */
+  init_() {
+    // TODO: Need the consider the case for shadow doc.
+    if (!this.isCookieAllowed_(this.win_, this.element_)) {
+      // Note: It's important to check origin here so that setCookie doesn't
+      // throw error "should not attempt ot set cookie on proxy origin"
+      return Promise.resolve();
+    }
+
+    if (!hasOwn(this.config_, 'cookies')) {
+      return Promise.resolve();
+    }
+
+    if (!isObject(this.config_['cookies'])) {
+      user().error(TAG, 'cookies config must be an object');
+      return Promise.resolve();
+    }
+
+    const inputConfig = this.config_['cookies'];
+
+    if (inputConfig['enabled'] === false) {
```

Comment:
Can this be a falsy value? If so change to `if (!inputConfig['enabled'])`

Reply:
`enabled` is true by default, so devs have to explicitly set `enabled` to false.